Allow pathfinding from multiple sources to a single destination.

dabreegster opened the pull request:
Hi @easbar,

I have a strange new use case for CHs, and as far as I can tell, this very simple change seems to work. I want to calculate the shortest path between multiple sources and a single destination. The resulting path should use the closest source. (I'm not trying to calculate one path from each source, but rather let multiple sources all be possibilities for the result.)

The reason for this: I model directed road segments as nodes, and turns in intersections as edges. When a route begins at a building, currently vehicles always turn one direction onto the road connected to the driveway -- right in the US, left in the UK. So the pathfinding starts from just one node. But I want to try allowing vehicles to turn left or right out of a driveway -- in many cases, this will produce a much more realistic path.

I looked at how the start node is used in `calc_path`, and it looks extremely straightforward to generalize -- so I did. And in my initial tests, it works -- on my end, I'm hacking every request to use two start nodes, one of them at a fixed position on the map. I have a UI where I can click one start point and visualize the path to a destination I hover over. With this experimental code hooked up, the path goes from the clicked start point, until I cross the threshold where my fixed starting point is closer to the destination.

However, I don't have a theoretical understanding of how the forwards+backwards search happens in CHs. So although the code seems to work, I don't know if it even _should_, or if I'm somehow getting lucky in my experiment so far.

If this kind of change is algorithmically valid, then I'll clean up the API, add docs, and add a unit test.

And maybe later, I'll investigate the same idea for multiple end points -- to model entering the destination driveway from either side of the road.

Thanks!

easbar commented:
Yes something like this is definitely possible. I guess what you are implicitly doing is adding a helper node that will be your start node and that is connected with zero weight edges to all your starting nodes. I mean this query

```
  start1
  start2              target
  start3
```
can be transformed to a single source query like this:

```
        / start1
start --- start2           target
        \ start3
```
where all the start-startx edges have zero weight (more generally they could even be weighted so different starting nodes could be favored relative to the others). After calculating the shortest path from start to target all we would need to do is checking whether the path took start1,2 or 3.

To be most confident about the correctness of your changed algorithm I think it is best to run random queries on generated graphs similar to here: https://github.com/easbar/fast_paths/blob/0d4fa89c9b503dd076e59cf617344443c72ba321/src/lib.rs#L126-L132
Checking the correctness should be simple because we can just calculate all the startx-target paths using our current algorithm (or use dijkstra as baseline) and check if the new one yields the minimum.

easbar commented:
I think we could generalize the API of the fast path algorithm such that it accepts multiple start and target nodes and maybe even add initial weights to all of them and provide the current API using a wrapper function that simply uses a single source and target without initial weight. Or would this be too confusing as some users might expect this returns all shortest paths between all sources and targets instead of the shortest path out of all possible combinations of sources and targets?

dabreegster commented:
Thank you for explaining -- your diagram with the zero-weight edges to multiple starts makes sense, and that was my intuition behind what should work.

The tests you suggested sound perfect, I'll work on those.

> I think we could generalize the API of the fast path algorithm such that it accepts multiple start and target nodes and maybe even add initial weights to all of them

I like the sound of this. I tried making a similar change for multiple target nodes, but I've hit a problem here: https://github.com/easbar/fast_paths/blob/0d4fa89c9b503dd076e59cf617344443c72ba321/src/path_calculator.rs#L225

`extract_nodes` just takes the original `target` node and pushes it onto the end. Is there a way to recover it from the forwards/backwards search?

> provide the current API using a wrapper function that simply uses a single source and target without initial weight

That's my plan. If we change the signature of `calc_path`, that's a breaking change warranting a new semver version. I'll make a new method accepting the more general start/ends.

> Or would this be too confusing as some users might expect this returns all shortest paths between all sources and targets instead of the shortest path out of all possible combinations of sources and targets

The API returns `Option<ShortestPath>`, so I don't think there'll be any expectation to get back multiple paths. The documentation can also be very clear. I'm not sure what terminology is best here. `calc_paths` is too vague. `calc_path_multiple_endpoints` maybe?

dabreegster committed:
Allow pathfinding from one of several sources to a single destination.

A new API is added for this, and the existing calls continue to work the
same.

dabreegster commented:
Alright, I cleaned up the method names, added docs, and added a new test. This is ready for review now.

No support yet for multiple target nodes, because I don't understand how to figure out which target the forwards/backwards search finds. If there's a way to recover that, then I can extend the API further.

easbar commented:
Thanks a lot for this and sorry for not getting back to you earlier. There are a few things I'd like to include with this new feature:

* allow multiple targets 
* document the new method in README
* use starting weights other than zero in the test

If you don't mind I'll carry on from here, I can do it this weekend.

easbar commented on src/path_calculator.rs:
```diff
@@ -57,25 +57,40 @@ impl PathCalculator {
-            return Some(ShortestPath::singular(start));
+        for (id, _) in &starts {
+            if *id == end {
+                return Some(ShortestPath::singular(*id));
```
I think here we should include the start weight. Say we route from node 3 with weight 6 to node 3 the result should be `nodes=[3], weight=6`, right?

easbar replied:
... I did

dabreegster commented:
Feel free to take over! I'll be back at a computer later this weekend and
can take a closer look. I never figured out how to reconstruct the target
node used, so that's why it's only multiple sources

On Sat, Jul 3, 2021, 2:12 PM Andi ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/path_calculator.rs
> <https://github.com/easbar/fast_paths/pull/29#discussion_r663409591>:
>
> >          assert!(end < self.num_nodes, "invalid end node");
>          self.heap_fwd.clear();
>          self.heap_bwd.clear();
>          self.valid_flags_fwd.invalidate_all();
>          self.valid_flags_bwd.invalidate_all();
> -        if start == end {
> -            return Some(ShortestPath::singular(start));
> +        for (id, _) in &starts {
> +            if *id == end {
> +                return Some(ShortestPath::singular(*id));
>
> I think here we should include the start weight. Say we route from node 3
> with weight 6 to node 3 the result should be nodes=[3], weight=6, right?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/easbar/fast_paths/pull/29#pullrequestreview-698619392>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAMWLF4Z2TKT662FSXK4G43TV54M7ANCNFSM46WHWI5Q>
> .
>

easbar commented on src/path_calculator.rs:
```diff
@@ -161,8 +176,9 @@ impl PathCalculator {
-            return Some(ShortestPath::new(start, end, best_weight, node_ids));
+            let node_ids = self.extract_nodes(graph, end, meeting_node);
+            let chosen_start = node_ids[0];
+            return Some(ShortestPath::new(chosen_start, end, best_weight, node_ids));
```
And should we not also include the start weight here? I guess it's a matter of convention. We could either think of the start weights as 'priorities' that can just be used to make it more/less likely for certain nodes to end up being the starting node of the shortest paths, or we could think of the start weights as actual weights that will be included in the returned path weight. I think I would prefer the second option, but maybe there is a reason you did not include it here?

dabreegster commented:
Including the start weight in the final cost does make sense. I've lost
context, I think there might've been one case where it was hard to preserve
the cost, so I was trying to uniformly exclude them. Maybe it was the no-op
case of start=end?

On Sat, Jul 3, 2021, 2:20 PM Andi ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/path_calculator.rs
> <https://github.com/easbar/fast_paths/pull/29#discussion_r663410254>:
>
> > @@ -161,8 +176,9 @@ impl PathCalculator {
>          if meeting_node == INVALID_NODE {
>              return None;
>          } else {
> -            let node_ids = self.extract_nodes(graph, start, end, meeting_node);
> -            return Some(ShortestPath::new(start, end, best_weight, node_ids));
> +            let node_ids = self.extract_nodes(graph, end, meeting_node);
> +            let chosen_start = node_ids[0];
> +            return Some(ShortestPath::new(chosen_start, end, best_weight, node_ids));
>
> And should we not also include the start weight here? I guess it's a
> matter of convention. We could either think of the start weights as
> 'priorities' that can just be used to make it more/less likely for certain
> nodes to end up being the starting node of the shortest paths, or we could
> think of the start weights as actual weights that will be included in the
> returned path weight. I think I would prefer the second option, but maybe
> there is a reason you did not include it here?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/easbar/fast_paths/pull/29#pullrequestreview-698619743>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAMWLF4L2FLG3YT3WKVFT6TTV55JRANCNFSM46WHWI5Q>
> .
>

easbar commented on src/lib.rs:
```diff
@@ -187,6 +202,71 @@ mod tests {
+                    .flat_map(|(source, _)| dijkstra.calc_path(&dijkstra_graph, *source, target))
+                    .collect();
+
+                // It's likely that there are multiple shortest paths, from different source nodes.
+                // Compare every Dijkstra's path with the one CH path.
```
We have to make sure that `path_fast` is as short as the shortest among `dijkstra_paths`. And yes there can be multiple elements in `dijkstra_paths` with the same (shortest) weight. In this case we have to make sure that `fast_path` equals one of them. Is this what you meant here?

easbar commented:
Closing here in favor of #30
